pysrc/bytewax/inputs.py

```diff
@@ -8,41 +8,104 @@
-multi-tasking, never blocking but yielding a bare `None` if there is
-no new input.
+    If you want to implement a custom connector, instead subclass one
+    of the specific input sub-types below in this module.
 
-Yields:
+    """
 
-    Either a new item and state, or `None`.
+    def __json__(self):
+        """This is used by the Bytewax platform internally and should
+        not be overridden.
 
-"""
+        """
+        return {
+            "type": type(self).__name__,
+        }
 
 
 # TODO: Add ABC superclass. It messes up pickling. We should get rid
 # of pickling...
-class PartitionedInput:
-    """An input source with a fixed number of independent partitions."""
+class StatefulSource:
+    """Input source that maintains state of its position."""
+
+    @abstractmethod
+    def next(self) -> Optional[Any]:
+        """Attempt to get the next input item.
+
+        This must participate in a kind of cooperative multi-tasking,
+        never blocking but returning a bare `None` if there is no item
+        to emit currently.
+
+        Returns:
+
+            An item or `None`.
+
+        Raises:
+
+            StopIteration: When the source is complete.
+
+        """
+        ...
+
+    @abstractmethod
+    def snapshot(self) -> Any:
+        """Snapshot the position of the next read of this source.
+
+        This will be returned to you via the `resume_state` parameter
+        of your input builder.
+
+        Be careful of "off by one" errors in resume state. This should
+        return a state that, when built into a source, resumes reading
+        _after the last read item item_, not the same item that
+        `next()` last returned.
+
+        This is guaranteed to never be called after `close()`.
+
+        Returns:
+
+            Resume state.
+
+        """
+        ...
+
+    def close(self) -> None:
+        """Do any cleanup on this source when the dataflow completes
+        on a finite input.
+
+        This is not guaranteed to be called. It will not be called
+        during a crash.
+
+        """
+        pass
+
+
+class PartitionedInput(Input):
+    """An input with a fixed number of independent partitions.
+
+    Will maintain the state of each source and re-build using it
+    during resume. If the source supports seeking, this input can
+    support exactly-once processing.
+
+    Each partition must contain unique data. If you re-read the same data
+    in multiple partitions, the dataflow will process these duplicate
+    items.
+
+    """
 
     @abstractmethod
     def list_parts(self) -> Set[str]:
-        """List all partitions for this input by a string key.
+        """List all partitions by a string key.
 
-        This must consistently return the same keys when called by any
-        worker in a cluster.
+        This must consistently return the same keys when called by all
+        workers in all executions.
 
         Keys must be unique within this dataflow step.
```

Comment:
We're going to need to do a pass over all the documentation and at some point reference "executions" publicly in a few places, as the concept can't fully be hidden.

Perhaps a better way of describing this is "It will be called once on each worker during each execution. It must consistently return the same keys every call."

I think we should add a Glossary page to our docs. But that's for a later pass.